### PR TITLE
Adds deployment ready abstraction

### PR DESCRIPTION
**Pull request by bharathappali**
This PR adds the deployment ready abstraction which loops the kubernetes service check and returns if we can proceed.

This PR is currently work in progress, The Experiment Trial object needs to be passed to the abstraction to get a better visibility of ongoing process check. 

Signed-off-by: bharathappali <abharath@redhat.com>

### CODE REVIEW

File: src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java [hunk 1]
```diff
@@ -159,32 +159,29 @@ public void startExperimentTrials() {
...
-                        long millisMinutes = TimeUnit.MINUTES.toMillis(deploymentIsReadyWithinMinute);
-                        int count = (int)millisMinutes/EMConstants.TimeConv.DEPLOYMENT_CHECK_INTERVAL_IF_READY_MILLIS;
-                        for (int j = count; j > 0 && !deploymentHandler.isDeploymentReady(); j--) {   //Parameterize sleep duration hardcoded for 120
-                            try {
-                                LOGGER.debug("Still deployment is not ready for ExpName {} trail No {}", this.experimentTrial.getExperimentName(), tracker);
-                                Thread.sleep(EMConstants.TimeConv.DEPLOYMENT_CHECK_INTERVAL_IF_READY_MILLIS);
-                            } catch (InterruptedException e) {
-                                e.printStackTrace();
-                            }
+                        EMUtil.DeploymentReadinessStatus deploymentReadinessStatus = deploymentHandler.isDeploymentReady();
```
Comment (dinogun) [inline src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java:164]: Need to check if deployment is needed and only then call initiateDeploy if needed. In the monitoring only usecase, we might not need to do any deployment at all

Comment (dinogun) [reply src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java:164]: This also needs to be further tied into `input.json`. The input.json validation phase needs to set a flag that deployment is needed for this trial. That flag can be used here to determine if `initiateDeploy` needs to be called.

**Review by dinogun — CHANGES_REQUESTED** [review]
Please see the changes needed in the comments above

File: src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java (hunk 1 above)
Comment (bharathappali) [reply src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java]: Raised #502 to make the flags check, limiting this PR scope to abstract out the deployment ready part

File: src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java [hunk 2]
```diff
@@ -159,32 +159,29 @@ public void startExperimentTrials() {
...
+                        // Proceeding to load check as deployment is successful
+                        EMUtil.LoadAvailabilityStatus loadAvailabilityStatus = emLoadInterceptor.isLoadAvailable(this.experimentTrial);
+                        switch (loadAvailabilityStatus) {
+                            case LOAD_AVAILABLE:
+                                // Proceed to collect metrics as load is available
+                                break;
+                            case LOAD_NOT_AVAILABLE:
+                                // Proceed to exit gracefully as load is not available
+                                break;
                         }
```
[the remark this answers is not in the document]
Comment (bharathappali) [reply src/main/java/com/autotune/experimentManager/core/ExperimentTrialHandler.java:184]: Thanks for the review @dinogun , Raised https://github.com/kruize/autotune/pull/502 to make the flags check, limiting this PR scope to abstract out the deployment ready part

**Comment by bharathappali** [comment]
[working_loop.log](https://github.com/kruize/autotune/files/9565594/working_loop.log)
@dinogun  I confirm that changes are working and they are not breaking the loop. Attaching the log for your reference.

docker image with the PR changes: `bharathappali/exman:deployment-is-ready-abstraction`

File: src/main/java/com/autotune/experimentManager/core/DeploymentHandler.java [hunk 3]
```diff
@@ -58,17 +61,30 @@ public void initiateDeploy() {
...
+                // Will be replaced by a exponential looper mechanism
+                Thread.sleep(EMUtil.timeToSleep(j, EMUtil.ThresholdIntervalType.LINEAR) * 1000);
+            }
         } catch (Exception e) {
             LOGGER.error(e.toString());
             LOGGER.error(e.getMessage(), e.getStackTrace().toString());
             e.printStackTrace();
         }
-        return running;
+        if (running) {
```
Comment (dinogun) [inline src/main/java/com/autotune/experimentManager/core/DeploymentHandler.java:85]: I think this needs to be `if (true == running) {` to keep it consistent with the usage in the loop above.

**Review by dinogun — APPROVED** [review]
LGTM

**Pull request merged**